File: sdk-dotnet/UsageClient.cs
using System.Net.Http;
using System.Text;
using System.Text.Json;

namespace Tollara;

public record UsageCallbackResult(
    bool Success,
    int HttpStatus,
    string HttpStatusText,
    string RequestUrl,
    string? ResponseBody = null,
    string? NetworkError = null);

public record UsageReportResponse(
    int ReportSchemaVersion,
    string? Status,
    string? Warning,
    string? UserId,
    string? ServiceId,
    string? BillingModelType,
    string? MeasurementType,
    string? UnitLabel,
    UsageBreakdown? Breakdown);

public static class UsageClient
{
    public const string DefaultUsagePathPrefix = "/api/usage";

    public static string BuildUsageReportUrl(string usageServiceUrl, string? usagePathPrefix)
    {
        var baseUrl = usageServiceUrl.TrimEnd('/');
        var p = string.IsNullOrWhiteSpace(usagePathPrefix) ? DefaultUsagePathPrefix : usagePathPrefix.Trim();
        if (!p.StartsWith('/')) p = "/" + p;
        p = p.TrimEnd('/');
        return $"{baseUrl}{p}/report";
    }

    public static Task<UsageCallbackResult> ReportProgressAsync(HttpClient http, string progressUrl, string requestId,
        string stage, int percentageComplete, string serviceSecret, CancellationToken ct = default) =>
        ReportProgressAsync(http, progressUrl, requestId, stage, percentageComplete, null, serviceSecret, ct);

    public static Task<UsageCallbackResult> ReportProgressAsync(HttpClient http, string progressUrl, string requestId,
        string stage, int percentageComplete, string? errorMessage, string serviceSecret, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object>
        {
            ["stage"] = stage,
            ["percentageComplete"] = percentageComplete,
            ["timestamp"] = DateTime.UtcNow.ToString("o"),
        };
        if (errorMessage != null) body["errorMessage"] = errorMessage;
        var bodyStr = JsonSerializer.Serialize(body);
        return PostSignedUsageCallbackAsync(http, progressUrl, bodyStr, serviceSecret, ct);
    }

    public static Task<UsageCallbackResult> ReportCompletionAsync(HttpClient http, string callbackUrl, string requestId,
        CompletionStatus status, string serviceSecret, CancellationToken ct = default) =>
        ReportCompletionAsync(http, callbackUrl, requestId, status, null, null, null, 0, serviceSecret, ct);

    public static Task<UsageCallbackResult> ReportCompletionAsync(HttpClient http, string callbackUrl, string requestId,
        CompletionStatus status, string? result, decimal units, string serviceSecret, CancellationToken ct = default) =>
        ReportCompletionAsync(http, callbackUrl, requestId, status, result, null, null, units, serviceSecret, ct);

    public static Task<UsageCallbackResult> ReportCompletionAsync(HttpClient http, string callbackUrl, string requestId,
        CompletionStatus status, string? result, string? resultUrl, string? contentType, decimal units, string serviceSecret, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object>
        {
            ["status"] = status.ToApiString(),
            ["timestamp"] = DateTime.UtcNow.ToString("o"),
            ["units"] = units,
        };
        if (result != null) body["result"] = result;
        if (resultUrl != null) body["resultUrl"] = resultUrl;
        if (contentType != null) body["contentType"] = contentType;
        var bodyStr = JsonSerializer.Serialize(body);
        return PostSignedUsageCallbackAsync(http, callbackUrl, bodyStr, serviceSecret, ct);
    }

    private static async Task<UsageCallbackResult> PostSignedUsageCallbackAsync(
        HttpClient http, string urlWithQuery, string bodyStr, string serviceSecret, CancellationToken ct)
    {
        var (baseUrl, timestamp) = ParseUrlParams(urlWithQuery);
        if (timestamp == null)
        {
            var statusText = string.IsNullOrEmpty(urlWithQuery)
                ? "Missing or invalid callback/progress URL"
                : "Missing timestamp query parameter in URL";
            return new UsageCallbackResult(false, 0, statusText, baseUrl);
        }

        try
        {
            var signature = Hmac.CalculateHmacWithTimestamp(bodyStr, timestamp, serviceSecret);
            var req = new HttpRequestMessage(HttpMethod.Post, baseUrl);
            req.Headers.TryAddWithoutValidation(TollaraHeaders.Signature, signature);
            req.Headers.TryAddWithoutValidation(TollaraHeaders.Timestamp, timestamp);
            req.Content = new StringContent(bodyStr, Encoding.UTF8, "application/json");
            var res = await http.SendAsync(req, ct);
            var responseBody = await res.Content.ReadAsStringAsync(ct);
            var statusText = res.IsSuccessStatusCode ? res.ReasonPhrase ?? "OK" : res.ReasonPhrase ?? $"HTTP {(int)res.StatusCode}";
            return new UsageCallbackResult(
                res.IsSuccessStatusCode,
                (int)res.StatusCode,
                statusText,
                baseUrl,
                string.IsNullOrEmpty(responseBody) ? null : responseBody);
        }
        catch (Exception ex)
        {
            return new UsageCallbackResult(false, 0, "Network error", baseUrl, NetworkError: ex.Message);
        }
    }

    /// <summary>Report usage using the SDK default API origin and usage path prefix (same as <see cref="TollaraClient"/>).</summary>
    public static Task<UsageReportResponse> ReportUsageAsync(HttpClient http, string userId, string serviceId, decimal unitsUsed,
        string serviceSecret, CancellationToken ct = default) =>
        ReportUsageAsync(http, TollaraClient.DefaultApiUrl, userId, serviceId, unitsUsed, serviceSecret, null, null, ct);

    /// <summary>Report usage against an explicit usage service origin (for custom or local stacks).</summary>
    public static Task<UsageReportResponse> ReportUsageAsync(HttpClient http, string usageServiceUrl,
        string userId, string serviceId, decimal unitsUsed, string serviceSecret, CancellationToken ct = default) =>
        ReportUsageAsync(http, usageServiceUrl, userId, serviceId, unitsUsed, serviceSecret, null, null, ct);

    public static Task<UsageReportResponse> ReportUsageAsync(HttpClient http, string usageServiceUrl,
        string userId, string serviceId, decimal unitsUsed, string serviceSecret, DateTime? timestamp, CancellationToken ct = default) =>
        ReportUsageAsync(http, usageServiceUrl, userId, serviceId, unitsUsed, serviceSecret, timestamp, null, ct);

    public static async Task<UsageReportResponse> ReportUsageAsync(HttpClient http, string usageServiceUrl,
        string userId, string serviceId, decimal unitsUsed, string serviceSecret, DateTime? timestamp, string? usagePathPrefix, CancellationToken ct = default)
    {
        var ts = (timestamp ?? DateTime.UtcNow).ToUniversalTime();
        var epochSec = new DateTimeOffset(ts).ToUnixTimeSeconds();
        var iso = ts.ToString("o");
        var bodyStr = JsonSerializer.Serialize(new { userId, serviceId, unitsUsed, timestamp = iso });
        var tsStr = epochSec.ToString();
        var signature = Hmac.CalculateHmacWithTimestamp(bodyStr, tsStr, serviceSecret);
        var reportUrl = BuildUsageReportUrl(usageServiceUrl, usagePathPrefix);
        var req = new HttpRequestMessage(HttpMethod.Post, reportUrl);
        req.Headers.TryAddWithoutValidation(TollaraHeaders.Signature, signature);
        req.Headers.TryAddWithoutValidation(TollaraHeaders.Timestamp, tsStr);
        req.Content = new StringContent(bodyStr, Encoding.UTF8, "application/json");
        var res = await http.SendAsync(req, ct);
        res.EnsureSuccessStatusCode();
        var json = await res.Content.ReadAsStringAsync(ct);
        return ParseUsageReportResponse(json);
    }

    internal static UsageReportResponse ParseUsageReportResponse(string json)
    {
        var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        UsageBreakdown? breakdown = null;
        if (root.TryGetProperty("breakdown", out var br) && br.ValueKind == JsonValueKind.Object)
        {
            breakdown = JsonSerializer.Deserialize<UsageBreakdown>(br.GetRawText(), new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true,
            });
        }
        return new UsageReportResponse(
            root.TryGetProperty("reportSchemaVersion", out var rsv) && rsv.TryGetInt32(out var rv) ? rv : 0,
            root.TryGetProperty("status", out var s) ? s.GetString() : null,
            root.TryGetProperty("warning", out var w) && w.ValueKind == JsonValueKind.String ? w.GetString() : null,
            ReadString(root, "userId"),
            ReadString(root, "serviceId"),
            ReadString(root, "billingModelType"),
            ReadString(root, "measurementType"),
            ReadString(root, "unitLabel"),
            breakdown
        );
    }

    private static string? ReadString(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var p)) return null;
        if (p.ValueKind == JsonValueKind.Null) return null;
        return p.ValueKind == JsonValueKind.String ? p.GetString() : null;
    }

    private static (string baseUrl, string? timestamp) ParseUrlParams(string? url)
    {
        if (url is null)
            return ("", null);
        var idx = url.IndexOf('?');
        if (idx < 0) return (url, null);
        var baseUrl = url[..idx];
        var query = url[(idx + 1)..];
        string? timestamp = null;
        foreach (var part in query.Split('&'))
        {
            var eq = part.IndexOf('=');
            if (eq <= 0) continue;
            var key = Uri.UnescapeDataString(part[..eq]);
            if (key == "timestamp") timestamp = part[(eq + 1)..];
        }
        return (baseUrl, timestamp);
    }
}

File: sdk-python/src/agentvend_service_sdk/usage_client.py
import json
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, Optional
from urllib.parse import parse_qs, urlparse

from .agentvend_headers import AgentVendHeaders
from .completion_status import CompletionStatus
from .hmac_utils import calculate_hmac_with_timestamp

DEFAULT_USAGE_PATH_PREFIX = "/api/usage"


def _usage_report_iso_and_epoch_sec(timestamp: Optional[float]) -> tuple[str, str]:
    """Body uses ISO-8601 instant; HMAC header uses Unix epoch seconds (spec §3)."""
    if timestamp is None:
        sec = int(time.time())
    elif timestamp > 1e11:
        sec = int(timestamp // 1000)
    else:
        sec = int(timestamp)
    iso = datetime.fromtimestamp(sec, tz=timezone.utc).isoformat().replace("+00:00", "Z")
    return iso, str(sec)


def _usage_report_url(base_url: str, usage_path_prefix: Optional[str]) -> str:
    base = base_url.rstrip("/")
    p = (usage_path_prefix or DEFAULT_USAGE_PATH_PREFIX).strip()
    if not p.startswith("/"):
        p = "/" + p
    p = p.rstrip("/")
    return f"{base}{p}/report"


@dataclass
class UsageReportResponse:
    status: Optional[str]
    warning: Optional[str]
    is_over_limit: bool
    remaining_requests_per_period: int
    remaining_time_units_per_period: Optional[float]
    remaining_spending_cap: Optional[float]
    overage_rate: Optional[float]


def _parse_url_params(url: str) -> tuple[str, Optional[str]]:
    parsed = urlparse(url)
    base = f"{parsed.scheme}://{parsed.netloc}{parsed.path}" if parsed.scheme else url.split("?")[0]
    if not parsed.query:
        return base, None
    qs = parse_qs(parsed.query)
    timestamp = (qs.get("timestamp") or [None])[0]
    return base, timestamp


def report_progress(
    progress_url: str,
    request_id: str,
    stage: str,
    percentage_complete: int,
    service_secret: str,
    error_message: Optional[str] = None,
    *,
    session: Optional["requests.Session"] = None,
) -> bool:
    """POST progress to usage service. Requires 'requests'."""
    try:
        import requests
    except ImportError:
        raise ImportError("report_progress requires 'requests'. pip install requests")
    base_url, timestamp = _parse_url_params(progress_url)
    if not timestamp:
        return False
    body = {"stage": stage, "percentageComplete": percentage_complete, "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")}
    if error_message is not None:
        body["errorMessage"] = error_message
    body_str = json.dumps(body, separators=(",", ":"))
    signature = calculate_hmac_with_timestamp(body_str, timestamp, service_secret)
    sess = session or __import__("requests").Session()
    resp = sess.post(
        base_url,
        data=body_str,
        headers={
            "Content-Type": "application/json",
            AgentVendHeaders.SIGNATURE: signature,
            AgentVendHeaders.TIMESTAMP: timestamp,
        },
    )
    return resp.ok


def report_completion(
    callback_url: str,
    request_id: str,
    status: CompletionStatus,
    service_secret: str,
    *,
    units: float = 0.0,
    session: Optional["requests.Session"] = None,
) -> bool:
    """POST completion with status and units only."""
    return report_completion_full(
        callback_url,
        request_id,
        status,
        service_secret,
        units=units,
        session=session,
    )


def report_completion_with_result(
    callback_url: str,
    request_id: str,
    status: CompletionStatus,
    service_secret: str,
    result: str,
    *,
    units: float = 0.0,
    session: Optional["requests.Session"] = None,
) -> bool:
    """POST completion with inline result text."""
    return report_completion_full(
        callback_url,
        request_id,
        status,
        service_secret,
        result=result,
        units=units,
        session=session,
    )


def report_completion_full(
    callback_url: str,
    request_id: str,
    status: CompletionStatus,
    service_secret: str,
    *,
    result: Optional[str] = None,
    result_url: Optional[str] = None,
    content_type: Optional[str] = None,
    units: float = 0.0,
    session: Optional["requests.Session"] = None,
) -> bool:
    try:
        import requests
    except ImportError:
        raise ImportError("report_completion_full requires 'requests'. pip install requests")
    base_url, timestamp = _parse_url_params(callback_url)
    if not timestamp:
        return False
    body = {
        "status": status.api_value,
        "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "units": units,
    }
    if result is not None:
        body["result"] = result
    if result_url is not None:
        body["resultUrl"] = result_url
    if content_type is not None:
        body["contentType"] = content_type
    body_str = json.dumps(body, separators=(",", ":"))
    signature = calculate_hmac_with_timestamp(body_str, timestamp, service_secret)
    sess = session or requests.Session()
    resp = sess.post(
        base_url,
        data=body_str,
        headers={
            "Content-Type": "application/json",
            AgentVendHeaders.SIGNATURE: signature,
            AgentVendHeaders.TIMESTAMP: timestamp,
        },
    )
    return resp.ok


def report_usage(
    base_url: str,
    user_id: str,
    service_id: str,
    units_used: float,
    service_secret: str,
    *,
    usage_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> UsageReportResponse:
    """Report usage with current time as timestamp."""
    return report_usage_at(
        base_url,
        user_id,
        service_id,
        units_used,
        service_secret,
        timestamp=None,
        usage_path_prefix=usage_path_prefix,
        session=session,
    )


def report_usage_at(
    base_url: str,
    user_id: str,
    service_id: str,
    units_used: float,
    service_secret: str,
    timestamp: Optional[float] = None,
    *,
    usage_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> UsageReportResponse:
    try:
        import requests
    except ImportError:
        raise ImportError("report_usage requires 'requests'. pip install requests")
    iso, ts_sec = _usage_report_iso_and_epoch_sec(timestamp)
    body = {"userId": user_id, "serviceId": service_id, "unitsUsed": units_used, "timestamp": iso}
    body_str = json.dumps(body, separators=(",", ":"))
    signature = calculate_hmac_with_timestamp(body_str, ts_sec, service_secret)
    url = _usage_report_url(base_url, usage_path_prefix)
    sess = session or requests.Session()
    resp = sess.post(
        url,
        data=body_str,
        headers={
            "Content-Type": "application/json",
            AgentVendHeaders.SIGNATURE: signature,
            AgentVendHeaders.TIMESTAMP: ts_sec,
        },
    )
    resp.raise_for_status()
    data = resp.json()
    return UsageReportResponse(
        status=data.get("status"),
        warning=data.get("warning"),
        is_over_limit=bool(data.get("isOverLimit")),
        remaining_requests_per_period=int(data.get("remainingRequestsPerPeriod", 0)),
        remaining_time_units_per_period=_opt_float(data.get("remainingTimeUnitsPerPeriod")),
        remaining_spending_cap=_opt_float(data.get("remainingSpendingCap")),
        overage_rate=_opt_float(data.get("overageRate")),
    )


def _opt_float(v: Any) -> Optional[float]:
    if v is None:
        return None
    if isinstance(v, (int, float)):
        return float(v)
    return None
